https://s3.amazonaws.com/clickhouse-test-reports/57358/5e9994c904af665a387c5af0801146ef13343e97/integration_tests__tsan__[6_6]/integration_run_parallel1_0.log
Sanitizer found a problem: DiskLocal uses background pool, but it gets destroyed in DB::ContextSharedPart::shutdown() 

```
E           Exception: Sanitizer assert found for instance ==================
E           WARNING: ThreadSanitizer: heap-use-after-free (pid=8)
E             Atomic read of size 1 at 0x7b4800014ad8 by main thread (mutexes: write M0, write M1):
E               #0 pthread_mutex_lock <null> (clickhouse+0x7249f6b) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #1 std::__1::__libcpp_mutex_lock[abi:v15000](pthread_mutex_t*) build_docker/./contrib/llvm-project/libcxx/include/__threading_support:304:10 (clickhouse+0x20063b99) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #2 std::__1::mutex::lock() build_docker/./contrib/llvm-project/libcxx/src/mutex.cpp:38:14 (clickhouse+0x20063b99)
E               #3 std::__1::lock_guard<std::__1::mutex>::lock_guard[abi:v15000](std::__1::mutex&) build_docker/./contrib/llvm-project/libcxx/include/__mutex_base:94:27 (clickhouse+0x16adb4fb) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #4 DB::BackgroundSchedulePool::cancelDelayedTask(std::__1::shared_ptr<DB::BackgroundSchedulePoolTaskInfo> const&, std::__1::lock_guard<std::__1::mutex>&) build_docker/./src/Core/BackgroundSchedulePool.cpp:269:25 (clickhouse+0x16adb4fb)
E               #5 DB::BackgroundSchedulePoolTaskInfo::deactivate() build_docker/./src/Core/BackgroundSchedulePool.cpp:59:14 (clickhouse+0x16adb4fb)
E               #6 DB::DiskLocalCheckThread::shutdown() build_docker/./src/Disks/DiskLocalCheckThread.cpp:69:11 (clickhouse+0x17166f84) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #7 DB::DiskLocal::shutdown() build_docker/./src/Disks/DiskLocal.cpp:523:23 (clickhouse+0x17159f8e) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #8 DB::Context::shutdown() build_docker/./src/Interpreters/Context.cpp:4078:19 (clickhouse+0x1739fa72) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #9 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_12::operator()() const build_docker/./programs/server/Server.cpp:1598:5 (clickhouse+0xf316296) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #10 BasicScopeGuard<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_12>::invoke() build_docker/./base/base/../base/scope_guard.h:99:9 (clickhouse+0xf316296)
E               #11 BasicScopeGuard<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_12>::~BasicScopeGuard() build_docker/./base/base/../base/scope_guard.h:48:26 (clickhouse+0xf316296)
E               #12 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) build_docker/./programs/server/Server.cpp:1976:1 (clickhouse+0xf30a488) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #13 Poco::Util::Application::run() build_docker/./base/poco/Util/src/Application.cpp:315:8 (clickhouse+0x1d6555be) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #14 DB::Server::run() build_docker/./programs/server/Server.cpp:396:25 (clickhouse+0xf2f5c21) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #15 Poco::Util::ServerApplication::run(int, char**) build_docker/./base/poco/Util/src/ServerApplication.cpp:131:9 (clickhouse+0x1d675234) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #16 mainEntryClickHouseServer(int, char**) build_docker/./programs/server/Server.cpp:202:20 (clickhouse+0xf2f2d83) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #17 main build_docker/./programs/main.cpp:505:12 (clickhouse+0x72d06a0) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E           
E             Previous write of size 8 at 0x7b4800014ad8 by main thread:
E               #0 operator delete(void*, unsigned long) <null> (clickhouse+0x72cee4f) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #1 std::__1::default_delete<DB::BackgroundSchedulePool>::operator()[abi:v15000](DB::BackgroundSchedulePool*) const build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48:5 (clickhouse+0x173b1f79) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #2 std::__1::unique_ptr<DB::BackgroundSchedulePool, std::__1::default_delete<DB::BackgroundSchedulePool>>::reset[abi:v15000](DB::BackgroundSchedulePool*) build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:305:7 (clickhouse+0x173b1f79)
E               #3 DB::ContextSharedPart::shutdown() build_docker/./src/Interpreters/Context.cpp:641:30 (clickhouse+0x173b1f79)
E               #4 DB::Context::shutdown() build_docker/./src/Interpreters/Context.cpp:4069:13 (clickhouse+0x1739f68f) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #5 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_12::operator()() const build_docker/./programs/server/Server.cpp:1598:5 (clickhouse+0xf316296) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #6 BasicScopeGuard<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_12>::invoke() build_docker/./base/base/../base/scope_guard.h:99:9 (clickhouse+0xf316296)
E               #7 BasicScopeGuard<DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&)::$_12>::~BasicScopeGuard() build_docker/./base/base/../base/scope_guard.h:48:26 (clickhouse+0xf316296)
E               #8 DB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) build_docker/./programs/server/Server.cpp:1976:1 (clickhouse+0xf30a488) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #9 Poco::Util::Application::run() build_docker/./base/poco/Util/src/Application.cpp:315:8 (clickhouse+0x1d6555be) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #10 DB::Server::run() build_docker/./programs/server/Server.cpp:396:25 (clickhouse+0xf2f5c21) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #11 Poco::Util::ServerApplication::run(int, char**) build_docker/./base/poco/Util/src/ServerApplication.cpp:131:9 (clickhouse+0x1d675234) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #12 mainEntryClickHouseServer(int, char**) build_docker/./programs/server/Server.cpp:202:20 (clickhouse+0xf2f2d83) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E               #13 main build_docker/./programs/main.cpp:505:12 (clickhouse+0x72d06a0) (BuildId: fab5336bbe3329137f7bd0a09c959cb58a8b6924)
E           
```

--- a/src/Interpreters/Context.cpp
+++ b/src/Interpreters/Context.cpp
@@ -588,6 +588,28 @@ struct ContextSharedPart : boost::noncopyable
             cache->cache->deactivateBackgroundOperations();
 
         {
+            // Disk selector might not be initialized if there was some error during
+            // its initialization. Don't try to initialize it again on shutdown.
+            if (merge_tree_disk_selector)
+            {
+                for (const auto & [disk_name, disk] : merge_tree_disk_selector->getDisksMap())
+                {
+                    LOG_INFO(log, "Shutdown disk {}", disk_name);
+                    disk->shutdown();
+                }
+            }
+
+            /// Special volumes might also use disks that require shutdown.
+            auto & tmp_data = root_temp_data_on_disk;
+            if (tmp_data && tmp_data->getVolume())
+            {
+                auto & disks = tmp_data->getVolume()->getDisks();
+                for (auto & disk : disks)
+                    disk->shutdown();
+            }
+        }
+
+        {
             std::lock_guard lock(mutex);
 
             /** Compiled expressions stored in cache need to be destroyed before destruction of static objects.
@@ -4066,26 +4088,6 @@ void Context::stopServers(const ServerType & server_type) const
 
 void Context::shutdown() TSA_NO_THREAD_SAFETY_ANALYSIS
 {
-    // Disk selector might not be initialized if there was some error during
-    // its initialization. Don't try to initialize it again on shutdown.
-    if (shared->merge_tree_disk_selector)
-    {
-        for (auto & [disk_name, disk] : getDisksMap())
-        {
-            LOG_INFO(shared->log, "Shutdown disk {}", disk_name);
-            disk->shutdown();
-        }
-    }
-
-    /// Special volumes might also use disks that require shutdown.
-    auto & tmp_data = shared->root_temp_data_on_disk;
-    if (tmp_data && tmp_data->getVolume())
-    {
-        auto & disks = tmp_data->getVolume()->getDisks();
-        for (auto & disk : disks)
-            disk->shutdown();
-    }
-
     shared->shutdown();
 }
 

